<h1>Q7) How many records have a new line?</h1>

In [15]:
import re
import pandas as pd
# Load the uploaded CSV file
file_path = 'Data.csv'
df = pd.read_csv(file_path)

# Display the first few rows and the column names to understand the structure
df.head(), df.columns

(                                                Data
 0  Watch or listen live weekdays at 8:30am MT at ...
 1  Watch or listen live weekdays at 8:30am MT at ...
 2               Chubby And Hot, Always Stir The Pot!
 3               Chubby And Hot, Always Stir The Pot!
 4  Journalist, publisher of Rebel News — telling ...,
 Index(['Data'], dtype='object'))

In [16]:
duplicate_count = df.duplicated().sum()
print("Number of completely duplicate rows:", duplicate_count)

Number of completely duplicate rows: 5353


In [17]:
text_duplicates = df['Data'].duplicated().sum()
print("Number of duplicate 'Data' entries:", text_duplicates)

Number of duplicate 'Data' entries: 5353


In [18]:
duplicated_rows = df[df['Data'].duplicated(keep=False)]
duplicated_rows.head()


,Data
0,Watch or listen live weekdays at 8:30am MT at ...
1,Watch or listen live weekdays at 8:30am MT at ...
2,"Chubby And Hot, Always Stir The Pot!"
3,"Chubby And Hot, Always Stir The Pot!"
6,Albertan. Mom to children & dogs. Wife. Friend...


In [19]:
df_no_duplicates = df.drop_duplicates(subset='Data')


In [20]:
df_no_duplicates = df_no_duplicates.reset_index(drop=True)


In [21]:
newline_pattern = r'\n'


In [26]:
mask = df_no_duplicates['Data'].str.contains(newline_pattern)
newline_count = mask.sum()
matching_rows = df_no_duplicates[mask]


In [27]:
print("✅ Number of records with a newline:", newline_count)
print(matching_rows[['Data']].head())

✅ Number of records with a newline: 299
                                                 Data
7   Think-tank president. Out-of-touch opinion eli...
13  Traveller, nature lover, photographer. HR Prof...
20  📌 Eloquent AF\n📌 Anti-Conservative\n📌 No time ...
24  When I look at the mess that is Alberta under ...
26  Alberta's Most Followed War Room--double the f...


**Factors to consider**
* Multiple newlines \n\n\n	May create tokenization artifacts or unintended paragraph breaks.
* Trailing/leading newlines	Can cause padding issues in prompt formatting.
* HTML <br> used instead	Missed by simple \n regex — should optionally detect.
* Invisibles like \r\n (Windows)	Could lead to false negatives if not normalized.
* Text split mid-sentence	Can affect context chunking or coherence in LLM summarization.


In [30]:
# Define a pattern to match all newline variants
newline_pattern = r'(?:\r\n|\r|\n|<br\s*/?>)'

# Count number of newlines in each record
df_no_duplicates['newline_count'] = df_no_duplicates['Data'].str.count(newline_pattern, flags=re.IGNORECASE)

# Count total records with at least one newline
has_newline_mask = df['newline_count'] >= 1
total_with_newlines = has_newline_mask.sum()

# Count total records with 2 or more newlines
multi_newline_mask = df_no_duplicates['newline_count'] >= 2
multi_newline_count = multi_newline_mask.sum()

# Print results
print(f"✅ Total records with at least one newline: {total_with_newlines}")
print(f"✅ Records with 2 or more newlines: {multi_newline_count}")

# Optional: preview those records
print("\n🔍 Sample of records with 2+ newlines:")
print(df_no_duplicates[multi_newline_mask][['Data', 'newline_count']].head())


✅ Total records with at least one newline: 299
✅ Records with 2 or more newlines: 184

🔍 Sample of records with 2+ newlines:
                                                 Data  newline_count
20  📌 Eloquent AF\n📌 Anti-Conservative\n📌 No time ...              4
24  When I look at the mess that is Alberta under ...              2
42  Ban semi auto rifles w/detachable magazines an...              2
61                                  Teacher\n\nHe/Him              2
73  Greek #Archaeology. State formation, #Thessaly...              2


<h2> 299 records contain a decimal number </h2>

**Reference List (APA 7):**
OpenAI. (2025). ChatGPT (May 17 version) [Large language model]. https://chat.openai.com/  

First Prompt: "How to use regex to find newlines"

Last Prompt : "count how many records have 2 or more newlines? That would help detect paragraphs or long posts, which often require chunking for LLM input"